### ISSUE TITLE
Input events are occasionally skipped (regression from input handling changes in 3.3)

### ISSUE THREAD

**Issue by saward** [godotengine/godot#48369]
___

***Bugsquad edit:** This issue has been confirmed several times already. No need to confirm it further.*

___

**Godot version:**
3.3

**OS/device including version:**

AMD Ryzen 3600
5700 XT
Arch Linux 5.11.16-arch1-1

**Issue description:**

With Godot 3.2.3, the tilemaps I'd set up performed poorly when zoomed out.  Now, with 3.3, it's  smooth which is great, but now it seems to be missing input events occasionally.

Most of the time using WASD to pan the camera works fine, but sometimes it keeps moving in a direction once I release key press, and sometimes it doesn't start moving immediately.  Outputting the events, it looks like the key press (or key release) isn't registered.  I don't recall noticing this in 3.2.3, but I did not test extensively.

**Steps to reproduce:**

Load and run the attached test project, full screen, zoomed out (should be by default), and try panning around with WASD.  Hopefully you observe that some inputs are being skipped -- that the panning isn't always moving exactly in line with the keys you're pressing

I tested also on a Windows machine with a weaker CPU and GPU, and the issue doesn't seem to be (as) apparent.

**Minimal reproduction project:**

[tilemap_camera.zip](https://github.com/godotengine/godot/files/6410577/tilemap_camera.zip)

**Comment by Calinou** [comment]
Can you reproduce this if you call `Input.set_use_accumulated_input(false)` in the main scene script's `_ready()` function?

**Comment by saward** [comment]
I noticed I'm using _physics_process instead of _process for camera inputs.  Tested these permutations:

- _phsyics_process & accumulate=default
- _physics_process & accumulated=false
- _process & accumulate=default
- _process & accumulate=false

While the problem is evident using all, it seemed a bit harder to trigger with Input.set_use_accumulated_input(false).  This isn't a rigorous experiment though so it may also be making no difference.

**Comment by saward** [comment]
I should add, most of the time (most inputs) it doesn't happen, but happens often enough for me to notice.

**Comment by jusw85** [comment]
I've had a very similar issue after updating to Godot 3.3, also on Arch Linux. I fixed it by disabling vsync in Project Settings. Let me know if it helps.

**Comment by jusw85** [comment]
I have no issue with 3.2.3, only with 3.3. Also, it seems to be input in general, not tied to the camera. 

Here's my reproduction project:

[test.zip](https://github.com/godotengine/godot/files/6411040/test.zip)

Try holding down right, then repeatedly tapping left. Sometimes the input isn't registered immediately.

After adding `Input.set_use_accumulated_input(false)`, it seems to help, but doesn't fix it entirely. Only disabling vsync solves it. This is standalone desktop executable, no issue with HTML5 export.

**Comment by Calinou** [comment]
@jusw85 Can you test this on all older 3.3 RCs and betas to determine when the regression started? You can download them [here](https://downloads.tuxfamily.org/godotengine/3.3/) and [here](https://downloads.tuxfamily.org/godotengine/3.2.4/) (3.2.4 was renamed to 3.3 during the development process).

**Comment by jusw85** [comment]
> @jusw85 Can you test this on all older 3.3 RCs and betas to determine when the regression started? You can download them [here](https://downloads.tuxfamily.org/godotengine/3.3/) and [here](https://downloads.tuxfamily.org/godotengine/3.2.4/) (3.2.4 was renamed to 3.3 during the development process).

Issue appears in beta 1 `Godot_v3.2.4-beta1_x11.64.zip`

No issue with 3.2.2 `Godot_v3.2.2-stable_x11.64.zip`

**Comment by saward** [comment]
Just thought I'd mention that I've just tried 3.3.2 and the issue is still present.

**Comment by Wolfe2x7** [comment]
In my opinion the title should be changed, because I have had a frustrating time with inputs in general, like @jusw85, and I think this is all related.

Under circumstances I cannot determine, making it difficult to reproduce, Godot may ignore inputs, period. With how my project works, this sometimes causes analog-input-based values to become stuck, which is game-breaking. It may be fixed simply by closing and restarting the game. I think Godot might also be reporting duplicate "is_action_just_pressed" events, but I'm not sure yet.

I've had enough and I am going to try disabling Vsync and capping the framerate at 144 to see if it helps ("debug/settings/fps/force_fps" is not only for debug mode, is it?). If that doesn't work, I intend to try going back to Godot 3.2.2, because the symptoms are unacceptable. :|

**Comment by Calinou** [comment]
> ("debug/settings/fps/force_fps" is not only for debug mode, is it?)

It works in exported projects too.

**Comment by Wolfe2x7** [comment]
> It works in exported projects too.

Good to know in general, thanks.

I will spend some time with 3.2.2, 3.2.3, and 3.2.4 to see if I can corroborate @jusw85's observation.

**Comment by jusw85** [comment]
`bisect` from 3.2.3-stable to 3.3-stable.

This is the commit which introduces the problem: f725d9cb733e103c91d889777b78c5a646095658

To clarify, I'm not missing any inputs. All my inputs are registered. Instead, my (keyboard) inputs are delayed and experiencing input lag. It doesn't happen consistently. Sometimes its very noticeable, other times barely noticeable. Disabling vsync in the project fixes the issue for me. I haven't tried other inputs such as mouse clicks. I estimate the input lag to be around 100ms - 500ms (definitely <1second)

I'm using Openbox on Arch Linux with the picom compositor with vsync enabled (on the compositor)

**Comment by Wolfe2x7** [comment]
I have booted my project many times in 3.2.3 stable now, and I cannot reproduce the symptom I've reported for 3.2.4 beta 1 and 3.3.2 stable. I have also tried turning off Vsync, which does not prevent the symptom. Turning off "Always on Top" has no effect either, so I don't think it's related to #48312.

Unfortunately I still have not been able to find a clue on what circumstances are causing the problem in the first place. When it happens, sometimes it can be repeated for a period of time each time the game is restarted, while sometimes restarting the game may fix it. When it isn't happening, I can't predict how many times I could restart the game before the problem appears again.

I've tried closing other programs when it happens, and I've tried deliberately acquiring keyboard focus in another window (like right here) to make it appear. Neither works.

**Comment by pouleyKetchoupp** [comment]
For anyone who an reproduce the issue, could you please test with #49720 to see if it helps?

There's a mutex to protect multi-threaded operations for events which I've tried to switch to a spin lock to see if it can avoid some random stalling.

**Comment by jusw85** [comment]
> For anyone who can reproduce the issue, could you please test with #49720 to see if it helps?
> 
> There's a mutex to protect multi-threaded operations for events which I've tried to switch to a spin lock to see if it can avoid some random stalling.

Doesn't help.

I simulated keypresses using xdotool to demonstrate the issue. It repeats right then left with 0.1s interval.
```
sleep 1; for i in {0..20}; do xdotool keydown Right; sleep 0.1; xdotool keyup Right; xdotool keydown Left; sleep 0.1; xdotool keyup Left; done
```
3.2.3

https://user-images.githubusercontent.com/3406652/122643592-2f35c880-d143-11eb-8ff9-997c15037de0.mp4

#49720, 3.3.*


https://user-images.githubusercontent.com/3406652/122643644-80de5300-d143-11eb-84de-2210f84b4bcc.mp4


https://user-images.githubusercontent.com/3406652/122643645-820f8000-d143-11eb-8f40-23f372d33787.mp4

**Comment by hunterloftis** [comment]
That excellent reproduction from @jusw85 is exactly what I experience with 3.3.2 on Ubuntu Desktop.

Despite all the improvements in 3.3, reliable input is a critical component for any game, so I'll have to stick with 3.2.3 for now.

**Comment by CritCorsac** [comment]
I did some testing on this issue and I've discovered something that might be of interest when running games with v-sync enabled in Godot 3.3.

When using Linux Mint 20.2, the stated input lag issue is very noticeable and the inputs will randomly delay themselves for almost up to a full second seemingly at random intervals. Regular gameplay becomes much harder, if not totally impossible in some cases because of this.
So I tried using Debian 10.10 (non-free) and found that the issue is still present, but it is much less invasive. It only really becomes noticeable when using the terminal command provided by @jusw85 a couple of times when running on Debian.

Both distros were run on the same machine, but I think it might be best if someone else could also confirm this discrepancy between the two distros on their own computer.

**Comment by Gnumaru** [comment]
I'm using godot 3.3.3 on ubuntu 21.04 X64 and I have both the default gnome environment and LXDE (openbox) installed.

Interestingly with me the vsync workaround is the oposite: when vsync is turned off I get the input delay, while with vsync turned on input works without problems. This happens both on gdm with xorg (gnome) and openbox with xorg (lxde). I havent tested with wayland. Input.set_use_accumulated_input(false) doesn't seem to make any difference.

I'm testing only keyboard input and I would like to notice that the problem seems to happen most commonly on the keyup events. The keydown events doesnt seem to have any problem, but the keyup events several times have a big delay of about one second between the actual mechanical release of the key and the engine noticing it.

The bug can be tested creating an empty scene, attaching the following script and running it on the editor:
```gdscript
extends Object
func _input(event):
	print(event.as_text())
func _physics_process(delta):
	print(Input.is_action_pressed('ui_right'))
```
While the scene is running, keep your eyes on the console (preferably a dedicated terminal window, not the built-in console) and do very quick key taps of the 'right arrow' key with a small interval of about half second between each keypress. One will note that sometimes the keydown and keyup events happen right after one other or with a very few number of frames between. But sometimes, the keyup will delayed by a large amount, like one second.

**Comment by pouleyKetchoupp** [comment]
Thanks for all the feedback and test scripts, it's going to be very useful! I'm sorry I didn't look into this issue in the past couple months, but I'll try to find some time soon to see if I can figure out what's going on.

**Comment by YaroslavFox** [comment]
I've recently faced the same problem. In my case it happens only when gamepad is connected. Otherwise I don't face this problem. And v-sync has nothing to do for me.
godot 3.3.2 windows 11

**Comment by Gnumaru** [comment]
> In my case it happens only when gamepad is connected. Otherwise I don't face this problem
Do you mean that, when a gamepad is connected, the problem occurs with gamepad input only? or with any other kind of key input, like if you connect the gamepad and start experiencing the problem also on keyboard input?

**Comment by hunterloftis** [comment]
> I've recently faced the same problem. In my case it happens only when gamepad is connected. Otherwise I don't face this problem. And v-sync has nothing to do for me.
godot 3.3.2 windows 11

Given that it's a different input altogether, vsync doesn't impact it, and it's on a wholly different operating system, I would be surprised if this were indeed the same issue.

**Comment by pouleyKetchoupp** [comment]
@jusw85 I'm able to reproduce the issue using your script, but for me it happens in both 3.2.3 and 3.3.2 whether vsync is on or off. I'm on ArchLinux / Gnome.
Is it always perfect on 3.2.3 like in the video for you, or do you experience similar issues,  just more consistently in 3.3.2? Does it happen in 3.2.3 under any circumstances?

**Comment by jusw85** [comment]
@pouleyKetchoupp

I tested again using 3.2.3 from [here](https://downloads.tuxfamily.org/godotengine/3.2.3/Godot_v3.2.3-stable_x11.64.zip) and 3.3 from [here](https://downloads.tuxfamily.org/godotengine/3.3/Godot_v3.3-stable_x11.64.zip). I'm on Arch / Openbox + picom (vsync on)

On 3.3, the issue almost always occurs (with vsync on). Sometimes it's less noticeable, sometimes more. This is one of the more noticeable ones. Note the occasional responsiveness issues with the input.

https://user-images.githubusercontent.com/3406652/132810138-adb1e788-13d1-4085-913f-34eb75b5fa18.mp4

On 3.2.3, I don't notice any problems. However, because I added a marker for the midpoint, I observed that sometimes it drifts away from the start (although sometimes it does). This however might be due to timing precision in `sleep 0.1` and having an extra frame of `right down` or 1 frame delay of `right up`.

https://user-images.githubusercontent.com/3406652/132810600-0fd0ee13-bd18-4788-8a21-e5a3c3f74604.mp4
